### PR TITLE
[FEATURE] IDLE tokens

### PR DESCRIPTION
**Pull request by buba-developer**
Added Idle Tokens for TopUp
Refactored multiplier logic of special tokens (with unwraps)

### CODE REVIEW

File: src/services/v2/api/coinGecko/CoinGeckoAPIService.ts [hunk 1]
```diff
@@ -79,6 +83,21 @@ export class CoinGeckoAPIService extends MultiChainAPIService {
...
+
+    for (let i = addresses.length - 1; i >= 0; i--) {
+      const address = addresses[i];
+      const cachedPrice = this.pricesCache.get(address);
+      if (
+        cachedPrice != undefined &&
+        cachedPrice.expirationTimestampMs > Date.now()
+      ) {
+        console.log(`Get price for token address ${addresses[i]}`);
+        addresses.splice(i, 1);
```
Comment (Sinu5oid) [inline src/services/v2/api/coinGecko/CoinGeckoAPIService.ts:96]: It's not a good idea to mutate the same array you are iterating
Let's rewrite it to use immutables

File: src/services/v2/api/coinGecko/CoinGeckoAPIService.ts [hunk 2]
```diff
@@ -197,6 +215,11 @@ export class CoinGeckoAPIService extends MultiChainAPIService {
         result[recoveredKey][currency] =
           data[assetAddress][currency].toString(10);
       }
+
+      this.pricesCache.set(recoveredKey, {
+        value: result[recoveredKey],
+        expirationTimestampMs: 5 * 60 * 1000
```
Comment (Sinu5oid) [inline src/services/v2/api/coinGecko/CoinGeckoAPIService.ts:221]: Based on the validity condition check, it seems like `Date.now()` is missing here

File: src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts [hunk 3]
```diff
@@ -76,7 +130,83 @@ export class DebitCardOnChainService extends MoverOnChainService {
...
+    wrappedAsset: SmallTokenInfo,
+    unwrappedAsset: SmallTokenInfo,
+    unwrappedAmount: string
+  ): Promise<WrappedData> {
+    const specialTokenHandler = this.specialTokenHandlers.find((h) =>
+      h.canHandle(wrappedAsset.address, this.network)
+    );
+
+    if (specialTokenHandler === undefined) {
+      // unwrappedAsset should be equal to wrappedAsset
```
Comment (Sinu5oid) [inline src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts:195]: Let's raise a Sentry exception here with a detailed breadcrumbs the moments before
Our just make an assertion that will throw an error if it fails

File: src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts [hunk 4]
```diff
@@ -317,4 +443,48 @@ export class DebitCardOnChainService extends MoverOnChainService {
...
+    );
+    if (
+      inWalletToken !== undefined &&
+      greaterThan(inWalletToken.priceUSD, '0')
+    ) {
+      return inWalletToken.priceUSD;
+    }
+
+    try {
+      const priceRecord =
```
Comment (Sinu5oid) [inline src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts:461]: You can also modify TheGraph service to serve as a fallback

Comment (Sinu5oid) [reply src/services/v2/on-chain/mover/debit-card/DebitCardOnChainService.ts:461]: And/or make a composition price service (see swap services as an example)

File: src/services/v2/on-chain/wrapped-tokens/WrappedToken.ts [hunk 5]
```diff
@@ -10,21 +10,40 @@ import { SmallToken, SmallTokenInfo } from '@/wallet/types';
 
 export abstract class WrappedToken extends PromiEventWrapper {
   readonly network: Network;
+  readonly web3: Web3;
+  readonly accountAddress: string;
 
-  constructor(network: Network) {
+  protected constructor(accountAddress: string, network: Network, web3: Web3) {
```
Comment (Sinu5oid) [inline src/services/v2/on-chain/wrapped-tokens/WrappedToken.ts:16]: Is there any reason to make a protected (available with inheritance only) constructor?
Keyword `abstract` prevents you from creating an instance of such classes by default

**Comment by buba-developer** [comment]
Already merged with #419

**Pull request closed without merging**